src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,358 @@
+    left_table = _right_in_table;
+    right_table = _left_in_table;
+
+    left_column_id = _right_column_id;
+    right_column_id = _left_column_id;
+  }
+
+  auto left_data_type = left_table->column_type(left_column_id);
+  auto right_data_type = right_table->column_type(right_column_id);
+
+  _pos_list_left = std::make_shared<PosList>();
+  _pos_list_right = std::make_shared<PosList>();
+
+  _is_outer_join = (_mode == JoinMode::Left || _mode == JoinMode::Right || _mode == JoinMode::Outer);
+
+  // Scan all chunks from left input
+  for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < left_table->chunk_count(); ++chunk_id_left) {
+    auto column_left = left_table->get_chunk(chunk_id_left).get_column(left_column_id);
+
+    // for Outer joins, remember matches on the left side
+    std::vector<bool> left_matches;
+
+    if (_is_outer_join) {
+      left_matches.resize(column_left->size());
+    }
+
+    // Scan all chunks for right input
+    for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < right_table->chunk_count(); ++chunk_id_right) {
+      auto column_right = right_table->get_chunk(chunk_id_right).get_column(right_column_id);
+
+      resolve_data_and_column_type(left_data_type, *column_left, [&](auto left_type, auto& typed_left_column) {
+        resolve_data_and_column_type(right_data_type, *column_right, [&](auto right_type, auto& typed_right_column) {
+          using LeftType = typename decltype(left_type)::type;
+          using RightType = typename decltype(right_type)::type;
+
+          // make sure that we do not compile invalid versions of these lambdas
+          constexpr auto left_is_string_column = (std::is_same<LeftType, std::string>{});
+          constexpr auto right_is_string_column = (std::is_same<RightType, std::string>{});
+
+          constexpr auto neither_is_string_column = !left_is_string_column && !right_is_string_column;
+          constexpr auto both_are_string_columns = left_is_string_column && right_is_string_column;
+
+          // clang-format off
+          if constexpr (neither_is_string_column || both_are_string_columns) {
+            auto iterable_left = create_iterable_from_column<LeftType>(typed_left_column);
+            auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
+
+            iterable_left.with_iterators([&](auto left_it, auto left_end) {
+              iterable_right.with_iterators([&](auto right_it, auto right_end) {
+                with_comparator(_scan_type, [&](auto comparator) {
+                  this->_join_two_columns(comparator, left_it, left_end, right_it, right_end, chunk_id_left,
+                                          chunk_id_right, left_matches);
+                });
+              });
+            });
+          }
+          // clang-format on
+        });
+      });
+    }
+
+    if (_is_outer_join) {
+      // add unmatched rows on the left for Left and Full Outer joins
```

Comment:
Seems to be the same discussion as above - we manually fall back to a NLJ implementation, so we "do" full outer. We probably just misunderstood how to do the fallback - let's clarify this tomorrow.

Reply:
Another point - we are not sure anymore why we don't want to support right and outer, since the `_right_matches` is a set of RowIDs, which we also obtain from the index.